### ISSUE TITLE
Allow operations to be seeded with Elements as well as ElementSeeds

### ISSUE THREAD

**Issue by p013570** [gchq/Gaffer#467]
ElementSeeds are just Elements with no group or properties. We could simplify operation chains and tidy up the GetAdjacentEntitySeeds operation if we could remove ElementSeeds.

Rather than removing we need to just change ElementSeed to be a parent of Element - or something similar.

**Comment by d21211122** [comment]
Does anyone have any objections to renaming ElementSeeds to ElementKeys? It looks a little odd if Element extends ElementSeed - could get confusing with ElementComponentKey, though?

**Comment by p013570** [comment]
Sounds good to me. Don't worry about the conflict with ElementComponentKey - there is another issue to remove ElementComponentKey - gh-90. I have started looking into that issue but may take a bit of time to finish it.

**Comment by p013570** [comment]
@d21211122 - do you mind if I finish this issue off?

**Comment by p013570** [comment]
I have been having a look into the seeds and ids and come up with the following:

```
ElementId
    -> EntityId
        -> EntitySeed
        -> Entity
    -> EdgeId
        -> EdgeSeed
        -> Edge
```

All operations now refer to IDs instead of seeds. This way you can seed an operation like GetElements with an EntitySeed or Entity.

**Comment by gaffer01** [comment]
If you seed the `GetElements` operation with an `Entity`, what results do I get back? e.g. If the seed/id is an `Entity` with say a property called `count` with a value of 10, do I only get back `Entity`s which match exactly, i.e. have exactly the same properties? (Presumably that is what should happen if the matching option is set to EQUALS.) If not, then it seems a bit confusing to allow `Entity`s to be seeds. If yes, then that requires quite a lot of changes to the current code as the operations would need to be changed to do the full matching.

**Comment by t616178** [comment]
Merged into develop-0.7.0

**Comment by p013570** [comment]
# Migration Steps

OK, this doesn't actually require a migration step but this feature does allow you to simplify some of your queries. GetElements operations can now be seeded with Elements.

For example if you previously had an operation like:
```java
final OperationChain opChain = new OperationChain.Builder()
                .first(new GetElements.Builder()
                        .input(seeds)
                        .view(new View.Builder()
                                .edge("red")
                                .build())
                        .build())
                .then(new GetElements.Builder()
                        .view(new View.Builder()
                                .entity("Cardinality", new ViewElementDefinition.Builder()
                                        .preAggregationFilter(new ElementFilter.Builder()
                                                .select("edgeGroup")
                                                .execute(new IsEqual(CollectionUtil.treeSet("blue")))
                                                .build())
                                        .groupBy()
                                        .postAggregationFilter(new ElementFilter.Builder()
                                                .select("hllp")
                                                .execute(new HyperLogLogPlusIsLessThan(5))
                                                .build())
                                        .build())
                                .build())
                        .build())
                .then(new GenerateObjects.Builder()
                        .generator(new EntitySeedExtractor())
                        .build())
                .then(new GetElements())
                .build();
```

You can remove the GenerateObjects step as the entities can be passed straight into the next GetElements operation.
